2D block load lowering for tt.dot operands with no intermediate op

LiyangLingIntel opened the pull request:
This PR is for https://github.com/intel/intel-xpu-backend-for-triton/issues/146
Based on discussion in https://github.com/intel/intel-xpu-backend-for-triton/pull/865, this is another version of implementation that we do not rely on the intermediate op `triton_intel_gpu.load_2d`. Instead, we have to make changes to `RewriteTensorPointer` pass. 
The strategy is to copy common `RewriteTensorPointer` pass to Intel GPU passes, do not rewrite `tt.load` op with TensorPoiner, as what the previous NV pass did. So that we can have it in later stage and lower it to `llvm.genx.GenISA.LSC2DBlockRead`.

chengjunlu commented on include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td:
```diff
@@ -185,4 +185,28 @@ def TritonIntelGPUMatchTargetSize : Pass<"tritonintelgpu-match-target-size", "ml
+def TritonIntelGPURewriteTensorPointer : Pass</*cli-arg*/"tritonintelgpu-rewrite-tensor-pointer", /*Op*/"mlir::ModuleOp"> {
+  let summary = "Rewrite load/stores with tensor pointers into legacy load/stores";
+  let description = [{
+    This pass rewrites all load/store semantics initiated by a `tt.make_tensor_ptr` and `tt.advance` into legacy
```
Change the comments for the things you made to support 2D load for now.

chengjunlu commented:
The RewriterTensorPointer.cpp is totally new in this PR. It is hard to review what is the change we made to it.

Please create a commit to copy the TTIR pass rewriter tensor pointer to the TT Intel GPU IR repo first. And then based on that, we can review the customized changes for Intel clearly.

LiyangLingIntel commented:
> The RewriterTensorPointer.cpp is totally new in this PR. It is hard to review what is the change we made to it.
> 
> Please create a commit to copy the TTIR pass rewriter tensor pointer to the TT Intel GPU IR repo first. And then based on that, we can review the customized changes for Intel clearly.

Rebased and adjusted the commit history, we can view this commit https://github.com/intel/intel-xpu-backend-for-triton/pull/941/commits/c01f7ac6373f2a67ccf2092be3343c40cbd953aa#diff-43a0aeab44c0c355cbd24ce57853a07b38d96b98d4d5b10a8a2e3dfbf121fdc4 to see the changes between Triton common `RewriteTensorPointer` pass and Intel `RewriteTensorPointer` pass.

etiotto commented on test/TritonIntelGPU/load-to-llvm-2dload.mlir:
```diff
@@ -0,0 +1,45 @@
+// RUN: triton-opt %s -split-input-file --intel-allocate-shared-memory --convert-triton-intel-gpu-to-llvm | FileCheck %s
```
Do we need ` --intel-allocate-shared-memory` to test the new pass. It would be better to do not add a dependency on that pass in a lit test designed to test only one transformation. Can you please remove it?

etiotto replied:
ok unfortunate. We *could* change the lit tests to have the changes `--intel-allocate-shared-memory` does but prob. not worth it.

etiotto commented on test/TritonIntelGPU/load-to-llvm-2dload.mlir:
```diff
@@ -0,0 +1,45 @@
+    %6 = tt.make_tensor_ptr %arg1, [%1, %4], [%5, %c1_i64], [%c0_i32, %c0_i32] {order = array<i32: 1, 0>} : <tensor<32x64xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #mma}>>>
+    %7 = tt.advance %3, [%c64_i32, %c-32_i32] : <tensor<64x32xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #mma}>>>
+    %8 = tt.advance %7, [%c-64_i32, %c32_i32] : <tensor<64x32xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #mma}>>>
+    // CHECK-COUNT-2: llvm.call @llvm.genx.GenISA.LSC2DBlockRead.v8i16({{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}, {{.*}}) : (i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32) -> vector<8xi16>
```
Given every argument is `{{.*}}` is not possible to determine the 2D block load is for one load etc...  Rewrite the CHECKS to provide more context.

LiyangLingIntel replied:
Changes are updated in https://github.com/intel/intel-xpu-backend-for-triton/pull/959

etiotto commented on test/TritonIntelGPU/load-to-llvm-2dload.mlir:
```diff
@@ -0,0 +1,45 @@
+// CHECK: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockRead.v8i32(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32) -> vector<8xi32>
+// CHECK: llvm.func spir_funccc @llvm.genx.GenISA.LSC2DBlockRead.v8i16(i64, i32, i32, i32, i32, i32, i32, i32, i32, i32, i1, i1, i32) -> vector<8xi16>
+#blocked = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [1, 16], warpsPerCTA = [2, 4], order = [1, 0]}>
+#mma = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [4, 2], A = [8, 16], B = [16, 16], C = [8, 16]}>
```
use #dpas not #mma

etiotto commented on test/TritonIntelGPU/load-to-llvm-2dload.mlir:
```diff
@@ -0,0 +1,45 @@
+#blocked = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [1, 16], warpsPerCTA = [2, 4], order = [1, 0]}>
+#mma = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [4, 2], A = [8, 16], B = [16, 16], C = [8, 16]}>
+module attributes {"triton_gpu.compute-capability" = 2 : i32, "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 16 : i32} {
+  tt.func public @matmul_no_scf_with_advance_kernel(%arg0: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg1: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg2: !tt.ptr<f32> {tt.divisibility = 16 : i32}, %arg3: i32 {tt.divisibility = 16 : i32}, %arg4: i32 {tt.divisibility = 16 : i32}, %arg5: i32 {tt.divisibility = 16 : i32}, %arg6: i32 {tt.divisibility = 16 : i32}, %arg7: i32 {tt.divisibility = 16  ...
```
Can we simplify the test by removing `{tt.divisibility = 16 : i32}` ? and `attributes {noinline = false}`. The goal of a good lit test is to be as small as possible (and still test that the pass works as  designed).

etiotto commented on test/TritonIntelGPU/tritonintelgpu-rewritetensorptr.mlir:
```diff
@@ -0,0 +1,105 @@
+// RUN: triton-opt %s -split-input-file --tritonintelgpu-rewrite-tensor-pointer=device-architecture=PVC | FileCheck %s
+
+// CHECK: #[[$BLOCKED:.+]] = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [1, 16], warpsPerCTA = [4, 16], order = [1, 0]}>
```
Add a `//COM: ...` to all the new tests to explain what are the objectives of the test (what does it test?)

etiotto commented on test/TritonIntelGPU/tritonintelgpu-rewritetensorptr.mlir:
```diff
@@ -0,0 +1,105 @@
+// CHECK: #[[$BLOCKED:.+]] = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [1, 16], warpsPerCTA = [4, 16], order = [1, 0]}>
+// CHECK: #[[$MMA:.+]] = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [16, 4], A = [8, 16], B = [16, 16], C = [8, 16]}>
+#blocked = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [1, 16], warpsPerCTA = [4, 16], order = [1, 0]}>
+#mma = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [16, 4], A = [8, 16], B = [16, 16], C = [8, 16]}>
```
#mma -> #dpas

whitneywhtsang replied:
done.

etiotto commented on test/TritonIntelGPU/tritonintelgpu-rewritetensorptr.mlir:
```diff
@@ -0,0 +1,105 @@
+    // CHECK:  tt.make_tensor_ptr {{.*}}, {{\[}}{{.*}}, {{.*}}], {{\[}}{{.*}}, {{.*}}], {{\[}}{{.*}}, {{.*}}] {order = array<i32: 1, 0>} : <tensor<32x256xf16, {{.*}}<{opIdx = 1, parent = #[[$MMA]]}>>>
+    %22 = tt.make_tensor_ptr %arg1, [%16, %20], [%21, %c1_i64], [%c0_i32, %19] {order = array<i32: 1, 0>} : <tensor<32x256xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #mma}>>>
+    %23:3 = scf.for %arg9 = %c0_i32 to %arg5 step %c32_i32 iter_args(%arg10 = %cst, %arg11 = %18, %arg12 = %22) -> (tensor<256x256xf32, #mma>, !tt.ptr<tensor<256x32xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #mma}>>>, !tt.ptr<tensor<32x256xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #mma}>>>)  : i32 {
+      // CHECK:  tt.load {{.*}} {boundaryCheck = array<i32: 0, 1>} : !tt.ptr<tensor<256x32xf16, {{.*}}<{opIdx = 0, parent = #[[$MMA]]}>>>
+      // CHECK:  tt.load {{.*}} {boundaryCheck = array<i32: 0, 1>} : !tt.ptr<tensor<32x256xf16, {{.*}}<{opIdx = 1, parent = #[[$MMA]]}>>>
```
This is the key part of this test right ? To test that the 2 `tt.load` are preserved ?

etiotto commented on test/TritonIntelGPU/tritonintelgpu-rewritetensorptr.mlir:
```diff
@@ -0,0 +1,105 @@
+      %34 = tt.advance %arg12, [%c32_i32, %c0_i32] : <tensor<32x256xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #mma}>>>
+      scf.yield %32, %33, %34 : tensor<256x256xf32, #mma>, !tt.ptr<tensor<256x32xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #mma}>>>, !tt.ptr<tensor<32x256xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #mma}>>>
+    }
+    // CHECK:  arith.truncf {{.*}}#0 : tensor<256x256xf32, #[[$MMA]]> to tensor<256x256xf16, #[[$MMA]]>
```
Are all of this `CHECKS` really necessary ?

etiotto commented on third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:
```diff
@@ -141,6 +142,182 @@ struct LoadOpConversion
+            elems[4], elems[5], elems[6]};
+  }
+
+  LogicalResult
```
Add a comment to indicate what this member function does. With an example.

etiotto commented on third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:
```diff
@@ -141,6 +142,182 @@ struct LoadOpConversion
+  LogicalResult
+  rewriteTensorPointerLoad(triton::LoadOp op, OpAdaptor adaptor,
+                           ConversionPatternRewriter &rewriter) const {
+    auto loc = op.getLoc();
+    auto typeConverter = getTypeConverter();
+    auto *ctx = rewriter.getContext();
```
Use the static type instead of `auto`.

etiotto commented on third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:
```diff
@@ -141,6 +142,182 @@ struct LoadOpConversion
+    Value other = op.getOther();
+
+    Type resultTy = op.getType();
+    if (auto tensorType = resultTy.dyn_cast<RankedTensorType>()) {
```
Given that the caller checks that the `ptr` argument of the load is a pointer to tensor, the result should of the load should be a tensor. So can you assert here (instead of an if...)

LiyangLingIntel replied:
Yes, indeed. I changed the `dyn_cast` to `cast`, which means here we assume `resultTy` is `RankedTensorType`. I noticed Triton upstream code does it in this way.

etiotto commented on third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:
```diff
@@ -141,6 +142,182 @@ struct LoadOpConversion
+
+    Type resultTy = op.getType();
+    if (auto tensorType = resultTy.dyn_cast<RankedTensorType>()) {
+      if (auto dotLayout =
```
Is better to avoid large nested regions in the code (helps readability). Can you early return if the tensor does not have `dot` layout ?

etiotto commented on third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:
```diff
@@ -141,6 +142,182 @@ struct LoadOpConversion
+        if (auto dpasLayout =
+                dotLayout.getParent().dyn_cast<DpasEncodingAttr>()) {
+
+          auto opIdx = dotLayout.getOpIdx();
```
This block of code is too large. Need to simplify/refactor this code.

etiotto commented on third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:
```diff
@@ -141,6 +142,182 @@ struct LoadOpConversion
+          int64_t elemsPerLane;
+          SmallVector<int64_t> elemsPerInstr;
+          if (opIdx == 0) {
+            auto shapeA = dpasLayout.getShapeA();
+            elemsPerInstr = {shapeA[0], shapeA[1]};
+            elemsPerLane = product<int64_t>(elemsPerInstr) /
+                           product<unsigned>(getThreadsPerWarp(dpasLayout));
+            unpackType = LLVM::getFixedVectorType(
+                typeConverter->convertType(eltTy), elemsPerLane);
+
+            // pack scalar to i16.
+            auto opsPerChannel = dpasLayout.getOpsPerChannel();
+            elemsPerLane = opsPerChannel == 4 ? elemsPerLane / 2 : elemsPerLane;
+            load2DGenXType =
+                LLVM::getFixedVectorType(type::i16Ty(ctx), elemsPerLane);
```
Can you try to factor out this code segment into a lambda so that you can pass the shape of either operand A or operand B and reuse the same code (the lambda) for both operands. 
Does this work?
```
auto getLoad2DType = [&](SmallVectorImpl<unsigned> &shape, IntegerType packedType){
    elemsPerInstr = {shape[0], shape[1]};
    int64_t elemsPerLane = product<int64_t>(elemsPerInstr) /
                                          product<unsigned>(getThreadsPerWarp(dpasLayout));    
     unpackType = LLVM::getFixedVectorType(
            typeConverter->convertType(eltTy), elemsPerLane);
            
      auto opsPerChannel = dpasLayout.getOpsPerChannel();

     elemsPerLane = (elemsPerLane / opsPerChannel);
     if (packedType.isInteger<16>) 
       elemsPerLane *= 2;

      load2DGenXType =
                LLVM::getFixedVectorType(packedType, elemsPerLane);
};
```

LiyangLingIntel replied:
Thanks for the suggestion, I have referred to your code and made corresponding changes in https://github.com/intel/intel-xpu-backend-for-triton/pull/959

etiotto reviewed (commented):
We should have a PR to copy the `RewriteTensorPointer.cpp` file over to the intel directory (a NFC PR). Then rebase this one.

etiotto commented on test/TritonIntelGPU/rewrite-tensor-pointer.mlir:
```diff
@@ -1,4 +1,7 @@
 // RUN: triton-opt %s -tritonintelgpu-rewrite-tensor-pointer | FileCheck %s
+// FIXME
+// XFAIL: *
```
Why does this fail?

LiyangLingIntel replied:
Because this lit test is designed for common `RewriteTensorPointer` pass, the input IR does not have `layoutEncoding` attribute, which Intel GPU `RewriteTensorPointer` pass needs to know this attribute to make decision.
I have added an extra condition in https://github.com/intel/intel-xpu-backend-for-triton/pull/958, if no layout attribute, Intel `RewriteTensorPointer` pass will behave as the common one.

LiyangLingIntel commented:
Based on offline discussion, I separated this pull request to 2 for the better reviewing:
* https://github.com/intel/intel-xpu-backend-for-triton/pull/958
* https://github.com/intel/intel-xpu-backend-for-triton/pull/959

I have marked this to draft for now and will close it when conversation under this PR are all resolved.

chengjunlu commented on third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:
```diff
@@ -7,21 +7,90 @@
+  }
+
+  // HW 2D block read instruction only supports stride[-1]=1.
+  auto fastChangeStride = strides.back();
```
Please use the `strides[order[0]]` for general supporting the Triton multi-dimensional representation.

chengjunlu commented on third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:
```diff
@@ -7,21 +7,90 @@
+  auto tensorShape = tensorType.getShape();
+
+  auto isPitchDivisible = false;
+  if (strides.size() == 2) {
```
Please file a JIRA to track the supporting on the column major B matrix which is required for the flash attention.

chengjunlu commented on third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:
```diff
@@ -542,14 +709,64 @@ class TritonIntelGPURewriteTensorPointerPass
+
+    DenseSet<Value> valueToRemove;
+    mod.walk([&valueToRemove, this](Operation *op) {
+      if (auto makeTensorPtrOp = dyn_cast<tt::MakeTensorPtrOp>(op)) {
```
We can re-align the changes here. I made different changes to support partial rewrite tensor pointer.

LiyangLingIntel commented on third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:
```diff
@@ -66,20 +136,66 @@ struct RewritedInfo {
+    // The loop below implements this algorithm.
+    SmallVector<Attribute, 4> layouts;
+    layouts.resize(rank);
+    layouts[rank - 1] = layout;
```
Done. Changes are updated to https://github.com/intel/intel-xpu-backend-for-triton/pull/958.
